### PR TITLE
Blog: Maybe in λ-Calculus

### PR DESCRIPTION
**Pull request by bash**
To do:
* [x] Example for `bind`
* [x] `filter`
* [x] Mention Mark Seemann's posts about maybe

### CODE REVIEW

File: content/blog/2023-04-06-lc-maybe.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
++++
+title = "Maybe in λ-calculus"
+description = "Read along as I implement the Maybe type and some useful functions for working on Maybe values in λ-calculus."
+draft = true
+authors = ["Ruben"]
++++
+
+I'm currently taking a class on [Models of Computation]. If you're studying at the University of Zurich or ETHZ I highly recommend that you take this class :) You'll get to solve really cool puzzles each week as part of the class.
+
+My favourite model of computation so far is [Lambda Calculus]. A [functional programming enthusiast], I may be a bit biased though…
```
Comment (Mafii) [inline content/blog/2023-04-06-lc-maybe.md:9]: "As a" instead of "A"

File: content/blog/2023-04-06-lc-maybe.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+Going forward, I will assume that you have some basic knowledge of [Lambda Calculus].
+
+## What is a Maybe?
+A `Maybe` value represents an optional value. It can either be present (which we will call `Just`) or absent (which we will call `Nothing`).
+In some programming languages, such as [Rust], `Maybe` is called `Option` instead.
```
Comment (Mafii) [inline content/blog/2023-04-06-lc-maybe.md:15]: you could add that Java has the Optional type

File: content/blog/2023-04-06-lc-maybe.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+# Note that this is equivalent to a function with three arguments:
+λa b s.s a b
+# ... or three functions with one argument each:
+λa.λb.λs.s a b
+```
+
+Picking out the first or second value is easy: We call our tuple value with a function that takes two arguments and returns one and discards the other.
+
+So picking the first value out of a tuple would look something like this:
+```λ
```
Comment (Mafii) [inline content/blog/2023-04-06-lc-maybe.md:65]: you legend <3

File: content/blog/2023-04-06-lc-maybe.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+```
+
+This is the same as `Just (TIMES_TWO X)`. Yayy it worked ✨
+We could do a similar demonstration for `Nothing`, but I leave that as an exercise to you dear reader.
+
+<!-- TODO: filter (with example, use IS_EVEN as predicate) -->
+
+## Outlook
+One of my favourite properties of λ-calculus is the composable nature of expressions. \
+We could easily build more complicated structures (trees, linked lists, etc.) from the functions discussed in this post. I might write a follow-up post soon :)
```
Comment (Mafii) [inline content/blog/2023-04-06-lc-maybe.md:258]: I might write a follow-up post soon :) <- maybe don't say soon, that's a classic :) what about `I might write a follow-up to investigate that`

Comment (Mafii) [reply content/blog/2023-04-06-lc-maybe.md:258]: or smth like that, just not soon :P

File: content/blog/2023-04-06-lc-maybe.md (hunk 3 above)
Comment (bash) [reply content/blog/2023-04-06-lc-maybe.md:65]: Thaaanks :)

File: content/blog/2023-04-06-lc-maybe.md (hunk 2 above)
Comment (bash) [reply content/blog/2023-04-06-lc-maybe.md]: d0252b37a171199d73b3db7b1f485013e0e4bb4f

File: content/blog/2023-04-06-lc-maybe.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+We can then generalize this to an expression that creates a tuple:
+```λ
+λa b.λs.s a b
+# Note that this is equivalent to a function with three arguments:
+λa b s.s a b
+# ... or three functions with one argument each:
+λa.λb.λs.s a b
+```
+
+Picking out the first or second value is easy: We call our tuple value with a function that takes two arguments and returns one and discards the other.
```
[the remark this answers is not in the document]
Comment (bash) [reply content/blog/2023-04-06-lc-maybe.md]: f6527ba80b68c88e6d049b4a5b5ffe5136405fe8

File: content/blog/2023-04-06-lc-maybe.md (hunk 4 above)
Comment (bash) [reply content/blog/2023-04-06-lc-maybe.md]: 60c3a3104a98ce70a035ff3b901c382ecd081efe

File: content/blog/2023-04-06-lc-maybe.md (hunk 1 above)
Comment (bash) [reply content/blog/2023-04-06-lc-maybe.md]: eee5a9df53dde569f4761c42efb75652c155ad49

File: content/blog/2023-04-06-lc-maybe.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+We can then generalize this to an expression that creates a tuple:
+```λ
+λa b.λs.s a b
+# Note that this is equivalent to a function with three arguments:
+λa b s.s a b
+# ... or three functions with one argument each:
+λa.λb.λs.s a b
+```
+
+Picking out the first or second value is easy: We call our tuple value with a function that takes two arguments, returns one and discards the other.
```
Comment (bash) [inline content/blog/2023-04-06-lc-maybe.md:63]:
```suggestion
Picking out the first or second value is easy: We call our tuple value with a function that picks one of its two arguments.
```

File: content/blog/2023-04-08-lc-maybe.md
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:1]: https://hackage.haskell.org/package/base/docs/Data-Maybe.html

Comment (Mafii) [reply content/blog/2023-04-08-lc-maybe.md:1]: https://en.wikipedia.org/wiki/Church_encoding

File: content/blog/2023-04-08-lc-maybe.md [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+So picking the first value out of a tuple would look something like this:
+```λ
+#              ╭─ This is the "constructor" from above
+#       ───────┴──────
+     ((λa.λb.λs.s a b) X Y) (λa.λb.a)
+#                            ───┬───
+#                               ╰─  This function  picks out
+#                                   the first of two arguments
+->>β (λs.s X Y)            (λa.λb.a)
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:103]: no need for this alignment

File: content/blog/2023-04-08-lc-maybe.md [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+->>β (λa.λb.a) X Y
+->>β X
+```
+
+Picking out the second value works analogously.
+
+> ℹ️ And now for a word ~~from our sponsor~~ about syntax: I'm using `->>β` to indicate β-reduction and `->>δ` to indicate δ-reduction (replacing definitions).
+
+### Booleans
+Booleans are encoded as expressions that take two arguments: One that is called when the boolean is true and one that is called when the boolean is false:
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:113]: Booleans can be

File: content/blog/2023-04-08-lc-maybe.md [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+#        │         ╭─ true
+#     ───┴────   ──┴───
+     (λb.b X Y) (λt f.t)
+#          ┬ ┬
+#          │ ╰─ If false, return Y
+#          ╰─ If true, return X
+->>β (λt f.t) X Y
+->>β X
+
+# See, we get X, because the boolean was true.
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:145]: As you can see

File: content/blog/2023-04-08-lc-maybe.md [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+->>β (λj n.j X) incr 0
+->>β incr X
+
+# Similarly, if we pass in a Nothing value, we get 0:
+     (λm. m incr 0)(λj n. n)
+->>β (λj n. n) incr 0
+->>β 0
+```
+
+> 💭 With a bit of imagination, you could see our `Maybe` value as an implementation of the [Visitor pattern] modulo the separate `Visitor` type.
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:212]: I wouldn't use the word modulo. It's very niche. What about without?

File: content/blog/2023-04-08-lc-maybe.md [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+
+Writing out these match expressions gets tedious quickly and is not very expressive.
+To improve on this, we'll implement some useful functions for working with `Maybe` values.
+
+### `bind`
+`bind` takes a `Maybe` value and a `Maybe`-returning function. It applies the function only if the `Maybe` value is `Just`. Otherwise it returns `Nothing`.
+You can think of this as a "apply some operation on the value and flatten the `Maybe`s".
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:219]: flatten the <nested>

File: content/blog/2023-04-08-lc-maybe.md [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+     map (Just X) double
+#    ─┬─
+#     ╰────────────╮
+#     ─────────────┴──────────────
+->>δ (λm f.bind m (λx. Just (f x))) (Just X) double
+#          ─┬──
+#           │
+#           ╰──────╮
+#           ───────┴────────
+->>β (λm f.(λm f.m f Nothing) m (λx. Just (f x))) (Just X) double
```
Comment (Mafii) [inline content/blog/2023-04-08-lc-maybe.md:292]: must be delta

File: content/blog/2023-04-08-lc-maybe.md (hunk 10 above)
Comment (bash) [reply content/blog/2023-04-08-lc-maybe.md:212]: I like it :)

**Pull request merged**